src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs

```diff
@@ -20,81 +20,199 @@ internal class EventPullerService : BackgroundService
-        catch (Exception ex)
+
+        private async Task RejectEvents(CancellationToken cancellationToken)
         {
-            switch (ex)
+            while (!cancellationToken.IsCancellationRequested)
             {
-                case DomainEventTypeNotRegisteredException:
-                case CloudEventSerializationException:
-                case DomainEventHandlerNotRegisteredException:
-                    this._logger.EventWillBeRejected(cloudEvent.Id, cloudEvent.Type, ex);
-                    await RejectEvent(eventGridTopicSubscription, lockToken, stoppingToken).ConfigureAwait(false);
-                    break;
-                default:
-                    this._logger.EventWillBeReleased(cloudEvent.Id, cloudEvent.Type, ex);
-                    await ReleaseEvent(eventGridTopicSubscription, lockToken, stoppingToken).ConfigureAwait(false);
-                    break;
+                await this._rejectEventChannel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false);
+                await this._eventGridTopicSubscription.Client.RejectCloudEventsAsync(this._eventGridTopicSubscription.TopicName, this._eventGridTopicSubscription.SubscriptionName, ReadCurrentContent(this._rejectEventChannel), cancellationToken).ConfigureAwait(false);
             }
         }
-    }
 
-    private static async Task AcknowledgeEvent(EventGridTopicSubscription eventGridTopicSubscription, string lockToken, CancellationToken stoppingToken)
-    { 
-        await eventGridTopicSubscription.Client.AcknowledgeCloudEventAsync(eventGridTopicSubscription.TopicName, eventGridTopicSubscription.SubscriptionName, lockToken, stoppingToken).ConfigureAwait(false);
-    }
+        private void SignalHandlerStarting()
+        {
+            Interlocked.Increment(ref this._handlersInProgress);
+        }
 
-    private static async Task ReleaseEvent(EventGridTopicSubscription eventGridTopicSubscription, string lockToken, CancellationToken stoppingToken)
-    {
-        await eventGridTopicSubscription.Client.ReleaseCloudEventAsync(eventGridTopicSubscription.TopicName, eventGridTopicSubscription.SubscriptionName, lockToken, stoppingToken).ConfigureAwait(false);
-    }
+        private void SignalHandlerCompleted()
+        {
+            Interlocked.Decrement(ref this._handlersInProgress);
+            lock (this._eventGridTopicSubscription)
+            {
+                this._taskCompletionSource.TrySetResult();
+                this._taskCompletionSource = new TaskCompletionSource();
+            }
+        }
 
-    private static async Task RejectEvent(EventGridTopicSubscription eventGridTopicSubscription, string lockToken, CancellationToken stoppingToken)
-    {
-        await eventGridTopicSubscription.Client.RejectCloudEventAsync(eventGridTopicSubscription.TopicName, eventGridTopicSubscription.SubscriptionName, lockToken, stoppingToken).ConfigureAwait(false);
+        private Task WaitForAvailableHandler(CancellationToken cancellationToken)
```

Comment:
Maybe this can be a `ValueTask` to reduce allocation in the best case?

If this is a bottleneck, which I'm not really not sure, a potential optimization would be to `IValueTaskSource` in addition to the `ValueTask`